base/ryu/Ryu.jl

```diff
@@ -0,0 +1,95 @@
+module Ryu
+
+include("utils.jl")
+include("shortest.jl")
+include("fixed.jl")
+include("exp.jl")
+
+neededdigits(::Type{Float64}) = 309 + 17
+neededdigits(::Type{Float32}) = 39 + 9
+neededdigits(::Type{Float16}) = 9 + 5
+
+"""
+    Ryu.writeshortest(x, plus=false, space=false, hash=true, precision=-1, expchar=UInt8('e'), padexp=false, decchar=UInt8('.'))
+    Ryu.writeshortest(buf::Vector{UInt8}, pos::Int, x, args...)
+
+Convert a float value `x` into its "shortest" decimal string, which can be parsed back to the same value.
+This function allows achieving the `%g` printf format.
+Note the 2nd method allows passing in a byte buffer and position directly; callers must ensure the buffer has sufficient room to hold the entire decimal string.
+
+Various options for the output format include:
+  * `plus`: for positive `x`, prefix decimal string with a `'+'` character
+  * `space`: for positive `x`, prefix decimal string with a `' '` character; overridden if `plus=true`
+  * `hash`: whether the decimal point should be written, even if no additional digits are needed for precision
+  * `precision`: minimum number of significant digits to be included in the decimal string; extra `'0'` characters will be added for padding if necessary
+  * `expchar`: character to use exponent component in scientific notation
+  * `padexp`: whether two digits should always be written, even for single-digit exponents (e.g. `e+1` becomes `e+01`)
+  * `decchar`: decimal point character to be used
+"""
+function writeshortest(x::T,
+        plus::Bool=false,
+        space::Bool=false,
+        hash::Bool=true,
+        precision::Integer=-1,
+        expchar::UInt8=UInt8('e'),
+        padexp::Bool=false,
+        decchar::UInt8=UInt8('.')) where {T <: Base.IEEEFloat}
+    buf = Vector{UInt8}(undef, neededdigits(T))
+    pos = writeshortest(buf, 1, x)
+    return unsafe_string(pointer(buf), pos-1)
```

Comment:
forcing an (unsafe) copy here seem unnecessary. above do `_string_n(neededdigits(T))` to make the Vector, then use `String(resize!(buf, pos - 1))` here

Reply:
What if we just add an `@assert` that `pos <= length(buf)`? It seems like that would actually ensure more safety (in the case `pos` becomes something unsafe-ly large)

Reply:
Yes, but that would still be slow. I'm suggesting you don't make an extra copy of the data here by sticking with the safe API.

Reply:
Oh.......got it. Yeah, we can do that.

Reply:
Well, we actually want `StringVector(n)` instead of `_string_n`, right? The underlying functions all expect a `Vector{UInt8}`.